server/blob/blob-union/src/main/java/org/apache/james/blob/union/HybridBlobStore.java

```diff
@@ -128,43 +146,59 @@ private HybridBlobStore(BlobStore lowCostBlobStore, BlobStore highPerformanceBlo
     @Override
     public Mono<BlobId> save(BucketName bucketName, InputStream data, StoragePolicy storagePolicy) {
         Preconditions.checkNotNull(data);
 
-        BufferedInputStream bufferedInputStream = new BufferedInputStream(data, configuration.getSizeThreshold() + 1);
-        return selectBlobStore(storagePolicy, Mono.fromCallable(() -> isItABigStream(bufferedInputStream)))
-            .flatMap(blobStore -> blobStore.save(bucketName, bufferedInputStream, storagePolicy));
+
+        Supplier<byte[]> byteSupplier = () -> {
+            try {
+                return readAllBytes(new BufferedInputStream(data, configuration.getSizeThreshold() + 1));
+            } catch (IOException e) {
+                LOGGER.error("Error when reading bytes from InputStream, cause: {}", e.getMessage());
+            }
+            return new byte[0];
+        };
+
+        return save(bucketName, byteSupplier.get(), storagePolicy);
     }
 
-    private Mono<BlobStore> selectBlobStore(StoragePolicy storagePolicy, Mono<Boolean> largeData) {
+    private Publisher<BlobStore> selectBlobStores(StoragePolicy storagePolicy, Mono<Boolean> largeData) {
         switch (storagePolicy) {
             case LOW_COST:
                 return Mono.just(lowCostBlobStore);
+
             case SIZE_BASED:
-                return largeData.map(isLarge -> {
-                    if (isLarge) {
-                        return lowCostBlobStore;
-                    }
-                    return highPerformanceBlobStore;
-                });
+                return largeData.flux()
+                    .filter(Boolean::booleanValue)
+                    .map(ignored -> lowCostBlobStore)
+                    .switchIfEmpty(
+                        Flux.just(configuration.isDuplicateWrite())
+                            .filter(duplicateWrite -> !duplicateWrite)
+                            .map(ignored -> highPerformanceBlobStore)
+                            .switchIfEmpty(Flux.just(highPerformanceBlobStore, lowCostBlobStore)));
+
             case HIGH_PERFORMANCE:
-                return Mono.just(highPerformanceBlobStore);
+                return Flux.just(this.configuration.isDuplicateWrite())
+                    .filter(duplicateWrite -> !duplicateWrite)
+                    .map(ignored -> highPerformanceBlobStore)
+                    .switchIfEmpty(Flux.just(highPerformanceBlobStore, lowCostBlobStore));
+
             default:
                 throw new RuntimeException("Unknown storage policy: " + storagePolicy);
         }
     }
 
-    private boolean isItABigStream(InputStream bufferedData) throws IOException {
-        bufferedData.mark(0);
-        bufferedData.skip(configuration.getSizeThreshold());
-        boolean isItABigStream = bufferedData.read() != -1;
-        bufferedData.reset();
-        return isItABigStream;
+    private byte[] readAllBytes(InputStream inputStream) throws IOException {
+        byte[] bytes = new byte[inputStream.available()];
```

Comment:
Well, that's working well in your tests because we rely on memory backed input streams. However, you can't assume all the stream to always be available.

 - in this case, we will alter the stream content - discarding not yet available bytes
 - Also we want to avoid an in-memory copy of large streams.

We will need to find a better strategy here. I suggest re-reading pervious comments I did above (temp files, getting rid of PERFORMING store policy, etc..)

Also, I'm available to do pair on this tomorrow.